processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -137,14 +137,12 @@ public BeanMappingMethod build() {
                 return null;
             }
 
-            if ( !method.getMappingOptions().isRestrictToDefinedMappings() ) {
```

Comment:
You are also right 😄. I think that we need to distinguish between composite (multiple) source parameters and single source parameters. For single source parameters, we need to ignore it, otherwise apply it. 

I think we can only change how we define the `MappingOptions`, because at the time we created them we know whether there are multiple source parameters or not. Something like `forceUpdateMethod`